rollup/internal/controller/sender/sender.go

```diff
@@ -446,141 +384,148 @@ func (s *Sender) resubmitTransaction(feeData *FeeData, auth *bind.TransactOpts,
+				return
+			}
+			if status == types.TxStatusFailed {
+				log.Warn("transaction already marked as failed, skipping resubmission", "hash", tx.Hash().String())
+				return
+			}
+
 			log.Info("resubmit transaction",
-				"hash", pending.tx.Hash().String(),
-				"from", pending.signer.From.String(),
-				"nonce", pending.tx.Nonce(),
-				"submit block number", pending.submitAt,
-				"current block number", number,
+				"hash", tx.Hash().String(),
+				"from", s.auth.From.String(),
+				"nonce", tx.Nonce(),
+				"submit block number", txnToCheck.SubmitBlockNumber,
+				"current block number", blockNumber,
 				"configured escalateBlocks", s.config.EscalateBlocks)
 
-			if tx, err := s.resubmitTransaction(pending.feeData, pending.signer, pending.tx); err != nil {
-				// If account pool is empty, it will try again in next loop.
-				if !errors.Is(err, ErrNoAvailableAccount) {
-					log.Error("failed to resubmit transaction, reset submitAt", "tx hash", pending.tx.Hash().String(), "err", err)
-				}
-				// This means one of the old transactions is confirmed
-				// One scenario is
-				//   1. Initially, we replace the tx three times and submit it to local node.
-				//      Currently, we keep the last tx hash in the memory.
-				//   2. Other node packed the 2-nd tx or 3-rd tx, and the local node has received the block now.
-				//   3. When we resubmit the 4-th tx, we got a nonce error.
-				//   4. We need to check the status of 3-rd tx stored in our memory
-				//     4.1 If the 3-rd tx is packed, we got a receipt and 3-nd is marked as confirmed.
-				//     4.2 If the 2-nd tx is packed, we got nothing from `TransactionReceipt` call. Since we
-				//         cannot do  anything about, we just log some information. In this case, the caller
-				//         of `sender.SendTransaction` should write extra code to handle the situation.
-				// Another scenario is private key leaking and someone send a transaction with the same nonce.
-				// We need to stop the program and manually handle the situation.
-				if strings.Contains(err.Error(), "nonce") {
-					// This key can be deleted
-					s.pendingTxs.Remove(key)
-					// Try get receipt by the latest replaced tx hash
-					receipt, err := s.client.TransactionReceipt(s.ctx, pending.tx.Hash())
-					if (err == nil) && (receipt != nil) {
-						// send confirm message
-						s.confirmCh <- &Confirmation{
-							ID:           pending.id,
-							IsSuccessful: receipt.Status == types.ReceiptStatusSuccessful,
-							TxHash:       pending.tx.Hash(),
-						}
-					} else {
-						// The receipt can be nil since the confirmed transaction may not be the latest one.
-						// We just ignore it, the caller of the sender pool should handle this situation.
-						log.Warn("Pending transaction is confirmed by one of the replaced transactions", "key", key, "signer", pending.signer.From, "nonce", pending.tx.Nonce())
+			if newTx, err := s.resubmitTransaction(s.auth, tx, baseFee); err != nil {
```

Comment:
one possible error is that: `one of the old transactions is confirmed` and in such case we should mark the transaction confirmed. But it rarely happens, we can leave it manually.

See the comments of the original function for more context.

Reply:
Yes. I use the following logic to handle this case in the loop:
1. If the tx is` confirmed`, update current tx's status as `confirmed`, and the sender's other txs (of the same nonce) as `failed`.
```
					// Update the status of the transaction to TxStatusConfirmed.
					if err := s.pendingTransactionOrm.UpdatePendingTransactionStatusByTxHash(s.ctx, tx.Hash().String(), types.TxStatusConfirmed, dbTX); err != nil {
						log.Error("failed to update transaction status by tx hash", "hash", tx.Hash().String(), "sender meta", s.getSenderMeta(), "from", s.auth.From.String(), "nonce", tx.Nonce(), "err", err)
						return err
					}
					// Update other transactions with the same nonce and sender address as failed.
					if err := s.pendingTransactionOrm.UpdateOtherTransactionsAsFailedByNonce(s.ctx, txnToCheck.SenderAddress, txnToCheck.Nonce, txnToCheck.Hash, dbTX); err != nil {
						log.Error("failed to update other transactions as failed by nonce", "senderAddress", txnToCheck.SenderAddress, "nonce", txnToCheck.Nonce, "excludedTxHash", txnToCheck.Hash, "err", err)
						return err
					}
```
2. When resubmitting txs, check if the "pending" tx has been marked `failed` because it may be confirmed by other "replaced" txs within the same loop.
```
			// It's possible that the pending transaction was marked as failed earlier in this loop (e.g., if one of its replacements has already been confirmed).
			// Therefore, we fetch the current transaction status again for accuracy before proceeding.
			status, err := s.pendingTransactionOrm.GetTxStatusByTxHash(s.ctx, tx.Hash().String())
			if err != nil {
				log.Error("failed to get transaction status by tx hash", "hash", tx.Hash().String(), "err", err)
				return
			}
			if status == types.TxStatusFailed {
				log.Warn("transaction already marked as failed, skipping resubmission", "hash", tx.Hash().String())
				return
			}
```

Reply:
seems that the above logic still has some problems... I need to separate the loop into **`two loops`**, one for confirm checking, and the other for resubmitting.